derp/derp_server.go

```diff
@@ -944,6 +1024,48 @@ func (s *Server) notePeerSendLocked(src key.NodePublic, dst *sclient) {
 	m[dst.key] = dst.connNum
 }
 
+// allocFlowStats allocates a flowStats to track bytes/packets sent in
+// the flow identified by this dstKey and srcOrFwd. It should only be
+// called when the flowStat for this connection does not yet exist.
+func (c *sclient) allocFlowStats(dstKey, srcOrFwd key.NodePublic) *flowStats {
+	c.debugLogf("allocating flow")
+	s := c.s
+
+	s.mu.Lock()
+	if _, ok := s.flows[c][dstKey]; !ok {
+		s.flows[c][dstKey] = make(map[key.NodePublic]*flowStats)
+	}
+	f := &flowStats{}
+	s.flows[c][dstKey][srcOrFwd] = f
+	s.mu.Unlock()
+
+	if _, ok := c.flows[dstKey]; !ok {
+		c.flows[dstKey] = make(map[key.NodePublic]*flowStats)
+	}
+	c.flows[dstKey][srcOrFwd] = f
+	return f
+}
+
+// updateFlowStats updates the flowStat tracking the data sent from
+// this sclient to the destination with dstKey. If this sclient is
+// forwarding the data from another mesh peer, then srcOrFwd is the
+// source key. If this sclient is forwarding the data to another mesh
+// peer, then srcOrFwd is the key of the mesh peer we are sending it
+// to. Otherwise, srcOrFwd is the zero value.
+func (c *sclient) updateFlowStats(dstKey, srcOrFwd key.NodePublic, contents []byte) {
+	var f *flowStats
+	// The common case is that the flowStat already exists
+	if _, ok := c.flows[dstKey]; ok {
+		f = c.flows[dstKey][srcOrFwd]
+	}
+	if f == nil {
+		f = c.allocFlowStats(dstKey, srcOrFwd)
+	}
+	f.packets += 1
+	f.bytes += int64(len(contents))
```

Comment:
so these are incremented without holding any locks, but then you read/clone them in the snapshot without any locks too? that looks racy.

If you don't have tests that are failing in race mode, there probably isn't sufficient new test coverage.

(or I'm not understanding the locking yet)